wireless/bluetooth/bt_netdev.c

```diff
@@ -945,7 +993,79 @@ static int btnet_req_data(FAR struct radio_driver_s *netdev,
       NETDEV_TXDONE(&priv->bd_dev.r_dev);
     }
 
-  UNUSED(priv);
+  return OK;
+}
+
+/****************************************************************************
+ * Name: btnet_req_hci_data
+ *
+ * Description:
+ *   Requests the transfer of a list of HCI frames to the MAC.
+ *
+ * Input Parameters:
+ *   priv      - Bluetooth network device
+ *   btmeta    - Bluetooth frame metadata
+ *   framelist - Head of a list of HCI frames to be transferred.
+ *
+ * Returned Value:
+ *   Zero (OK) returned on success; a negated errno value is returned on
+ *   any failure.
+ *
+ ****************************************************************************/
+
+static int  btnet_req_hci_data(FAR struct btnet_driver_s *priv,
+                               FAR struct bluetooth_frame_meta_s *meta,
+                               FAR struct iob_s *framelist)
+{
+  FAR struct iob_s *iob;
+  FAR struct bt_buf_s *buf;
+
+  /* Add the incoming list of frames to the MAC's outgoing queue */
+
+  for (iob = framelist; iob != NULL; iob = framelist)
+    {
+      /* Increment statistics */
+
+      NETDEV_TXPACKETS(&priv->bd_dev.r_dev);
+
+      /* Remove the IOB from the queue */
+
+      framelist     = iob->io_flink;
+      iob->io_flink = NULL;
+
+      DEBUGASSERT(iob->io_offset == BLUETOOTH_MAX_HDRLEN &&
+                  iob->io_len >= BLUETOOTH_MAX_HDRLEN);
+
+      /* Allocate a buffer to contain the IOB */
+
+      switch (iob->io_data[iob->io_offset])
+        {
+          case BT_H4_CMD:
+            iob->io_offset += 1;
+            buf = bt_buf_alloc(BT_CMD, iob, 0);
+            break;
+          case BT_H4_EVT:
+            iob->io_offset += 1;
+            buf = bt_buf_alloc(BT_EVT, iob, 0);
+            break;
+          default:
+            return -EOPNOTSUPP;
+            break;
+        }
+
+      if (buf == NULL)
+        {
+          wlerr("ERROR:  Failed to allocate buffer container\n");
+          return -ENOMEM;
+        }
+
+      g_btdev.btdev->send(g_btdev.btdev, buf);
```

Comment:
This should use `bt_send`

Reply:
This was initially like that. bt_send is actually an internal function so I exposed it as public. But I realized this is part of hcicore.c, and this socket would be skipping NuttX's host layer, so I think it is preferable to call this directly. Also, hcicore is the Intel owned code, just to be aware of.

Reply:
But we need a way to intercept all of the outgoing HCI transfers before they go to the driver to be captured by monitoring channels.  Also it is wrong to be calling `g_btdev` everywhere so I am trying to cut down on that to make supporting multiple devs easier later.

Reply:
Maybe this can be implemented in a new function in bt_netdev.c then. Not sure how to handle the multiple-device case yet.

Reply:
My plan for multiple devices was to add a field to the buf structure that contains the btdev pointer.

Reply:
This seems wrong to me why are we not putting the data in the tx_queue and having `hci_tx_kthread` do the sending.

Reply:
That was one of my questions earlier in the other PR: does it make sense to have two queues? I understand the HCI queue was probably there because of how it was integrated to userspace via direct calls. But once the networking layer is the only access point to HCI from userspace, do we really need a queue besides the networking one?

Reply:
> This would just be implementing the function you need in `bt_hcicore.c` that looks like `static void hci_cmd_status(FAR struct bt_buf_s *buf)` whith the logic that you need
Sorry, don't understand what you mean

Reply:
You still need the kernel to be able to manage the interface (I get you are bypassing it right now). If you send a scan for n sec control request via the management interface the kernel will need to handle things like making sure the stop scan is scheduled correctly.

Reply:
Maybe we need to extract the queue handling from hcicore.c into a different file then. Right now it is a mix of part of the host stack implementation and an API to communicate with the controller. For example, there's both a bt_initialize and hci_initialize there. If you agree I can do that. Sadly this means that even when using nimBLE we would still have Intel owned code built, but I guess we can worry about that later.

Reply:
> > This would just be implementing the function you need in `bt_hcicore.c` that looks like `static void hci_cmd_status(FAR struct bt_buf_s *buf)` whith the logic that you need
> 
> Sorry, don't understand what you mean

That was the wrong function sorry.  We should not be calling the driver here we should be calling a higher level API. 
Can we just call `bt_hci_cmd_send`

Reply:
> Maybe we need to extract the queue handling from hcicore.c into a different file then. Right now it is a mix of part of the host stack implementation and an API to communicate with the controller. For example, there's both a bt_initialize and hci_initialize there. If you agree I can do that. Sadly this means that even when using nimBLE we would still have Intel owned code built, but I guess we can worry about that later.

I think we can worry about that later.  I think once the GATT logic is remove there will actually be very little left of the Intel code and what is left is interfaces that are clearly inspired from things like bluez.  Carrying a few files with the Intel BSD licence is not a deal breaker, we are going to end up having that anyway in other parts of the OS we just want to minimize it.

Reply:
> My hope would be that we can play with using the `nimBLE` stack where we just leverage the GATT layer on top of the L2CAP sockets, but I understand wanting to first just run it fully in userspace.

That will probably require quite some work on nimBLE. At that point we should try to collaborate with mynewt devs to see how they feel about the idea and maybe try to get it integrated.

Reply:
> That was the wrong function sorry. We should not be calling the driver here we should be calling a higher level API.
> Can we just call `bt_hci_cmd_send`

`bt_hci_cmd_send` just adds one level of indirection before sending to the queue and the problem is that it requires introspecting the packet since it expects the opcode already. I think it is better to replace bt_send with bt_queue_send, which is what bt_hci_cmd_send uses. I will see if I can separate bt_* functions from hci_* functions and have it separately built so that it is easier to deactivate NuttX's host layer.

Reply:
Mmm, bt_queue_send receives the queue as parameter. I think this requires a bit of API rework to not expose the queue.
I'll take note of these changes but I really want to get RX working now that TX is working. I want to test nimBLE itself aswell.